compiler/qsc/src/interpret/circuit_tests.rs

```diff
@@ -0,0 +1,804 @@
+
+    expect![[r"
+        q_0    ── H ──── ● ──── Z ──────────────────────────────
+        q_1    ───────── X ─────┼──────────── Z ────────────────
+        q_2    ── H ─────────── ● ──── H ─────┼───── M ─────────
+                                              │      ╘══════════
+        q_3    ── H ───────────────────────── ● ──── H ──── M ──
+                                                            ╘═══
+    "]]
+    .assert_eq(&circ.to_string());
+}
+
+#[test]
+fn operation_with_qubit_arrays() {
+    let mut interpreter = interpreter(
+        r"
+        namespace Test {
+            @EntryPoint()
+            operation Main() : Result[] { [] }
+
+            open Microsoft.Quantum.Measurement;
+            operation Test(q1: Qubit[], q2: Qubit[][], q3: Qubit[][][], q: Qubit) : Result[] {
+                for q in q1 {
+                    H(q);
+                }
+                for qs in q2 {
+                    for q in qs {
+                        X(q);
+                    }
+                }
+                for qss in q3 {
+                    for qs in qss {
+                        for q in qs {
+                            Y(q);
+                        }
+                    }
+                }
+                X(q);
+                MeasureEachZ(q1)
+            }
+        }",
+        Profile::Unrestricted,
+    );
+
+    let circ = interpreter
+        .circuit(CircuitEntryPoint::Operation("Test.Test".into()))
+        .expect("circuit generation should succeed");
+
+    expect![[r"
+        q_0    ── H ──── M ──
+                         ╘═══
+        q_1    ── H ──── M ──
+                         ╘═══
+        q_2    ── X ─────────
+        q_3    ── X ─────────
+        q_4    ── X ─────────
+        q_5    ── X ─────────
+        q_6    ── Y ─────────
+        q_7    ── Y ─────────
+        q_8    ── Y ─────────
+        q_9    ── Y ─────────
+        q_10   ── Y ─────────
+        q_11   ── Y ─────────
+        q_12   ── Y ─────────
+        q_13   ── Y ─────────
+        q_14   ── X ─────────
+    "]]
+    .assert_eq(&circ.to_string());
+}
+
+#[test]
+fn adjoint_operation() {
+    let mut interpreter = interpreter(
+        r"
+        namespace Test {
+            @EntryPoint()
+            operation Main() : Result[] { [] }
+
+            operation SWAP (q1 : Qubit, q2 : Qubit) : Unit
+                is Adj + Ctl {
+
+                body (...) {
+                    CNOT(q1, q2);
+                    CNOT(q2, q1);
+                    CNOT(q1, q2);
+                }
+
+                adjoint (...) {
+                    SWAP(q1, q2);
+                }
+
+                controlled (cs, ...) {
+                    CNOT(q1, q2);
+                    Controlled CNOT(cs, (q2, q1));
+                    CNOT(q1, q2);
+                }
+            }
+        }",
+        Profile::Unrestricted,
+    );
+
+    let circ = interpreter
+        .circuit(CircuitEntryPoint::Operation("Adjoint Test.SWAP".into()))
```

Comment:
I think it would be helpful to test the adjoint of an operation that is not effectively "adjoint self" such that the circuit diagram will look different for the two. You could have a custom operation that has `X` as its body and `Y` as it's adjoint, since technically nothing verifies that explicit adjoints are actually the inverse of their body implementation. That would make it very obvious that the circuit below is for the adjoint!

Reply:
Good point. I wanted a realistic example but didn't know enough to come up with one. I think I'll just have a very simple and obvious one instead.